### PR TITLE
feat!: atomic memory opcodes

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Add new ACIR opcodes doing atomic memory operation. They allow ACVM to perform stepwise solving for dynamic memory.

Resolves 

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: acir/src/circuit/opcodes.rs [hunk 1]
```diff
@@ -36,6 +36,44 @@ pub enum Opcode {
...
+        op: MemOp,
+    },
+    MemoryInit {
+        block_id: BlockId,
+        init: Vec<Witness>,
+    },
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum OpcodeEnum {
```
Comment (kevaundray) [inline acir/src/circuit/opcodes.rs:51]: Why is this needed?

File: acvm/src/pwg/block.rs [hunk 2]
```diff
@@ -30,50 +30,65 @@ impl BlockSolver {
         self.block_value.get(&index).copied()
     }
 
+    pub(crate) fn init(&mut self, init: &[Witness], initial_witness: &WitnessMap) {
```
Comment (kevaundray) [inline acvm/src/pwg/block.rs:33]: can you add a doc comment saying what this is doing and why it is needed?

File: acvm/src/compiler/transformers/fallback.rs [hunk 3]
```diff
@@ -32,6 +32,13 @@ impl FallbackTransformer {
                     acir_supported_opcodes.push(opcode);
                     continue;
                 }
+                Opcode::MemoryInit { .. } | Opcode::MemoryOp { .. } => {
+                    if !is_supported(&opcode) {
+                        return Err(CompileError::UnsupportedOpcode(opcode.to_enum()));
```
Comment (kevaundray) [inline acvm/src/compiler/transformers/fallback.rs:37]: I think this is a bit misleading, since we have Arithmetic as a variant and this should never return the arithmetic variant because Arithmetic is always supported

File: acir/src/circuit/opcodes.rs (hunk 1 above)
Comment (guipublic) [reply acir/src/circuit/opcodes.rs:51]: This is needed to indicate that an opcode is not supported inside a CompilerError. The Opcode has a too big size to be used directly.

File: acvm/src/compiler/transformers/fallback.rs (hunk 3 above)
Comment (guipublic) [reply acvm/src/compiler/transformers/fallback.rs]: I have renamed the enum and make sure it can only be called for these two opcodes.

File: acir/src/circuit/opcodes.rs (hunk 1 above)
Comment (guipublic) [reply acir/src/circuit/opcodes.rs]: I have renamed it, following the later comment.

**Comment by TomAFrench** [comment]
Marking this breaking as it will change the ACIR serialisation

**Review by kevaundray — APPROVED** [review]
